plugins/asus-hid/fu-asus-hid-device.c

```diff
@@ -0,0 +1,628 @@
+							 c0,
+							 NULL,
+							 FU_ASUS_HID_REPORT_ID_FLASHING,
+							 error))
+			return FALSE;
+
+		if (!fu_asus_hid_device_write_blocks(self, blocks, progress, error))
+			return FALSE;
+
+		/* ("D0") command */
+		// TODO: add D0 command checking here.
+		if (!fu_asus_hid_device_transfer_feature(device,
+							 d0,
+							 d0_res,
+							 FU_ASUS_HID_REPORT_ID_FLASHING,
+							 error))
+			return FALSE;
+
+		/* ("C3") command */
+		fu_struct_asus_flush_page_set_address(c3, page_offset);
+		if (!fu_asus_hid_device_transfer_feature(device,
+							 c3,
+							 NULL,
+							 FU_ASUS_HID_REPORT_ID_FLASHING,
+							 error))
+			return FALSE;
+
+		fu_progress_step_done(progress);
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_asus_hid_device_write_firmware(FuDevice *device,
+				  FuFirmware *firmware,
+				  FuProgress *progress,
+				  FwupdInstallFlags flags,
+				  GError **error)
+{
+	if ((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "upgrades have not yet been validated");
+		return FALSE;
+	}
+
+	/* TODO: why isn't this applying from probe() to bootloader in emulation case? */
+	fu_hid_device_set_interface(FU_HID_DEVICE(device), 0);
+
+	if (!fu_asus_hid_device_verify_ite_part(device, error))
+		return FALSE;
+
+	/* automatic backup should happen when FWUPD_DEVICE_FLAG_BACKUP_BEFORE_INSTALL is set */
+
+	/* ASUS tool uses a series of 2 C2 commands that appear to check bootloader integrity:
+	 * Start out by dumping the first 8k and seeing if all 8k matches in 1024 byte pages
+	 *
+	 * Use C2 command to set next offset?  Or to clear region.
+	 *
+	 * If dump didn't match, next offset will be 0x2000 and size will be right shifted by 10
+	 * If dump did match, next offset will be 0 and size will be right shifted and 8 subtracted.
+	 *
+	 * c2(offset, shifted_size)
+	 */
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+
+	if (!fu_asus_hid_device_write_data(FU_ASUS_HID_DEVICE(device), firmware, progress, error))
+		return FALSE;
+
+	/* sequence of D1 commands, presumably to verify result */
+
+	/* reset using attach */
+
+	return TRUE;
+}
+
+static gboolean
+fu_asus_hid_device_set_quirk_kv(FuDevice *device,
+				const gchar *key,
+				const gchar *value,
+				GError **error)
+{
+	FuAsusHidDevice *self = FU_ASUS_HID_DEVICE(device);
+
+	if (g_strcmp0(key, "AsusHidNumMcu") == 0) {
+		guint64 tmp;
+
+		if (!fu_strtoull(value, &tmp, 0, G_MAXUINT8, FU_INTEGER_BASE_AUTO, error))
+			return FALSE;
+		self->num_mcu = tmp;
+		return TRUE;
+	}
+
+	/* failed */
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "quirk key not supported");
+	return FALSE;
+}
+
+static void
+fu_asus_hid_device_init(FuAsusHidDevice *self)
+{
+	fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_NO_AUTO_REMOVE_CHILDREN);
```

Comment:
do you need the children to outlive the parent?

Reply:
Yup!  Specifically when doing the detach/attach dance the parent disappears.  Without those children you won't have all the details of what payload goes to what child device.